### PR TITLE
Win. Use `uv_key_t` to store current thread handle in TLS on Windows prior Vista.

### PR DESCRIPTION
**Pull request by moteus**
When we use LoadLibrary to load libuv library process gets AV when trying assign to `uv__current_thread`.

PS. To build it also needs comment `'_WIN32_WINNT=0x0600',` line in `uv.gyp`.

### CODE REVIEW

**Comment by moteus** [comment]
ping

**Comment by saghul** [comment]
> This is why I prefer link with dynamic library. In this case you know when you unload.

Well, it's not a matter of preference, we can't break people's code while fixing this.

> But because waiting on Sync objects in DllMain is not safe it also quite difficult to stop threads in this stage. So all threads should be stopped in user code (uv_loop_close?) and DllMain could release all sync objects.

uv_loop_close just closes one loop, there might be others running. Maybe we need some "uv_destroy" to destroy all global state created by libuv (or better, try to get rid of it if possible). However, we don't have a public "uv_init", it's done automatically, so this would look weird.

> But to do this in LibUV we should support thread safe container of all finalizers.
> May be we can just use one critical section and basic linked list to register callback with context?

That sounds way too generic, can you elaborate?

TBH, I don't know a clean way to solve this at this point.

**Comment by moteus** [comment]
I agree that make libuv uloadable is a difficult task and it needs separate PR.
So I remove dllmain from current PR. Is there reason to not merge this PR?

**Comment by saghul** [comment]
> Is there reason to not merge this PR?

@piscisaureus, WDYT?

File: src/win/thread.c [hunk 1]
```diff
@@ -117,7 +117,32 @@ void uv_once(uv_once_t* guard, void (*callback)(void)) {
...
+static volatile int initialized = 0;
+
+
+static void cleanup(void) {
+  if (initialized == 0)
+    return;
+
+  uv_key_delete(&uv__current_thread);
+
+  initialized = 0;
```
Comment (txdv) [inline src/win/thread.c]: is the initialized == 0 and initialized = 0 intended?

**Comment by piscisaureus** [comment]
Thanks - landed in https://github.com/libuv/libuv/compare/libuv:ebc50ae...libuv:a310248

I removed any cleanup related code. I'm open to supporting proper cleanup for all of libuv, but only if there is a decent strategy to do it.

**Pull request closed without merging**